Add plain multisig wallet example

athei opened the pull request:
This implements a plain multisig wallet with configurable owners and confirmation threshold. It is modeled after the popular [gnosis multisig wallet](https://github.com/gnosis/MultiSigWallet/blob/master/contracts/MultiSigWallet.sol). It allows executing arbitrary calls using the wallet origin to other contracts after receiving enough confirmations of the contract `owners`.

**Note**
The [polkadot-js/apps UI](https://polkadot.js.org/apps/) cannot be used to create the contract because of this issue: polkadot-js/api#1549. However, you need a dedicated DAPP for this contract anyways to have a good UX.

# Differences from the gnosis wallet

This contract stays close to the gnosis wallet. However, I took the liberty of changing some parts of the contract where I saw room for improvement. 

## State cleanup
The gnosis contract does not clean up historic state whatsoever. This mainly concerns the map that stores the confirmations for each owner and transaction. This seems reasonable when there is no state rent and the cleanup is expensive in computation.  My implementation does this somewhat expensive pruning in each call that obsoletes the state. So every call cleans up after itself.

## Execution is always a separate call
In the gnosis contract a confirmation triggers the execution when the confirmation threshold is reached. I opted for a more simpler approach where the execution must be triggered separately. This avoids cluttering the return type of the confirmation that otherwise had to return information about the possible execution.

## Avoid heavy `ensure_confirmed` computation
The gnosis contract needs to filter through the confirmations when determining if a transaction is confirmed. We avoid this singularity by adding a `confirmation_count` storage item that is kept synchronous with the current confirmations for each transaction. The trade off is the added complexity of doing this.

# TODO
This is a draft for a reason. There are some things left todo.

- [x] Documentation
- [x] State clearly that this contract is an example and not audited or ready for production
- [x] Fix `cargo contract generate-metadata`
- [x] Emit Events
- [x] Switch from `HashMap` to `BTreeMap`. ~Blocked by #334~
- [x] Tests

athei commented:
Can you please comment on if you think that `execute_transaction` is correct? I assumed that this allows calling another contract with the contract itself ("the wallet") as the origin. I am not sure if this is correct.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+                .map_err(|_| ())
+        }
+
+        fn internal_confirm(&mut self, id: TransactionId) {
```
needs better name, e.g. `confirm_transaction_by_caller`

Robbepop replied:
but also I'd like to keep internal methods more descriptive, e.g. adding this `confirmer: AccountId` to the params and invoking it with `self.env().caller()` on call-site instead of using `self.env().caller()` in the implementation. but this is details.

athei replied:
Done. Resolve if satisfied.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+            assert!(self.is_owner.contains_key(&self.env().caller()));
+        }
+
+        fn ensure_from_wallet(&self) {
```
also what about call stacks where in-between the caller is not the same `account_id`?
E.g. `multi-sig` calls contract `A` calls contract `multi-sig`, then this is not identifiable.

athei replied:
I do not see the problem. Then the second `multi-sig` sees `A` as the sender which is exactly what I want to happen. Also I do not see how the chosen name is not good, I want to ensure that specific functions call only be called by the contract.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        fn owner_index(&self, owner: &AccountId) -> u32 {
+            self.owners.iter().position(|x| *x == *owner).unwrap() as u32
```
use `expect` instead of `unwrap`

Robbepop replied:
might it makes sense for `self.owners` to be of type `storage::HashMap` or `storage::BTreeMap` instead? Why is it a `storage::Vec`?

athei replied:
Because I have to iterate over them in `clean_transaction_confirmations`. As matter of fact there is also a owner HashMap called `is_owner`. This is used to avoid iteration on the more common operations.

Robbepop replied:
still requires change to `expect` with proper message

athei replied:
It is. This snippet is just outdated.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+            for (id, _) in self.transactions.iter() {
+                if self.confirmations.remove(&(id, *owner)).is_some() {
+                    self.confirmation_count
+                        .mutate_with(&id, |count| *count -= 1);
```
Couldn't we simply do this after clean-up of owners and set it directly to `0`? Would be more gas efficient.

Robbepop replied:
I see this being a performance hazard in case of many owners and transactions due to double-loop. We have to make sure that this is as efficient as possible and won't be a trap for gas costs.

athei replied:
Are you sure you are commenting on the correct line? Is see no double loop here.

athei commented:
> All in all looks good but needs adjustments and definitely documentation and plenty of tests.

Yes there will be definitely documentation and thorough testing (as soon as the problem with the tests is resolved).

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+
+use ink_lang as ink;
+
+#[ink::contract(version = "0.1.0", env = MyEnv)]
```
It did not work outside of the module. I guess the user of this type will be generated inside of he module. Maybe with an absolute path it would work when the type is declared outside.

athei replied:
I am not sure how to progress with this item. Should I crate an issue for this? I do not find that behavior surprising. (I am trying to get all items to resolved).

Robbepop replied:
outdated comments from what I can tell. resolved

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+    use scale::Output;
+
+    type TransactionId = u32;
+    type MyEnv = env::DefaultEnvTypes;
```
Yes doing it that way feels awkward. The only place where it is set should be as a parameter to the proc macro.

athei replied:
https://github.com/paritytech/ink/issues/330

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+
+    #[ink(storage)]
+    struct MultisigPlain {
+        confirmations: storage::HashMap<(TransactionId, AccountId), ()>,
```
I think this is reasonable. Multi dimensional maps are not that exotic.

Robbepop replied:
This was more about the `()` as value type.

athei replied:
You are right. I did not read properly. This also reasonable :D

athei replied:
https://github.com/paritytech/ink/issues/332

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+    #[ink(storage)]
+    struct MultisigPlain {
+        confirmations: storage::HashMap<(TransactionId, AccountId), ()>,
+        confirmation_count: storage::HashMap<TransactionId, u32>,
```
I guess this is only for the time being? Because as an example the `HashMap` is the more favorable structure in this case, right?

Robbepop replied:
Generally yes but until we don't know how to implement an efficient non-attackable `storage::HashMap` with our current `storage` data structure design we should advocate users to use `storage::BTreeMap` for all cases where normally `storage::HashMap` would better be used.

athei replied:
https://github.com/paritytech/ink/issues/334

athei replied:
I switched to `BTreeMap`

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+            for owner in owners.iter() {
+                self.is_owner.insert(*owner, ());
+                self.owners.push(*owner);
+            }
```
Seems like a good idea. Loops are ugly.

athei replied:
https://github.com/paritytech/ink/issues/333

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        fn ensure_transaction_exists(&self, id: TransactionId) {
+            self.transactions.get(id).unwrap();
```
Are you sure about that? Basically what we are assuming here is: "Okay the user gave us an invalid id. We do not care: Just rollback". So there is no real proof behind that that this cannot happen. It is totally plausible that this happens. It's not like in runtime where we actually `expect` it to not happen.

Robbepop replied:
Well that's what you would write into `expect` to make clear that panicking upon this situation is actually the way this API works. Otherwise contract devs looking at this code might think that this is a logic error etc.

athei replied:
That seems reasonable. I will remove all unwraps.

athei replied:
Done. Reopen if necessary.

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+        }
+    }
+
+    fn ensure_requirement(owners: u32, required: u32) {
```
Done.

riusricardo commented:
Just a heads-up. You need to implement a few more traits for `Transaction` to build the metadata. 
`cargo contract generate-metadata` fails to build.

athei commented:
Yeah I noticed this already. Thanks for pointing that out. I will add that to the TODO of this PR.

athei commented:
Yeah I noticed this already. Thanks for pointing that out. I will add that to the TODO of this PR.

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,231 @@
+        }
+
+        fn ensure_confirmed(&self, id: TransactionId) {
+            assert!(self.confirmation_count.get(&id).unwrap() >= self.required.get());
```
Done. Reopen if necessary.

athei commented:
Does anyone understand why the `examples-fmt` CI job fails? I cannot find a helpful message in the output.

Robbepop commented:
> 
> 
> Does anyone understand why the `examples-fmt` CI job fails? I cannot find a helpful message in the output.

```
  //! ```0 < requirement && requirement <= owners && owners <= MAX_OWNERS```
-//!
+//! 
```
Is the output. So it is unhappy about some double empty lines or some whitespace that follows after a new doc-line. Also I would definitely put the contents of the ````` into another line, e.g.

athei commented:
You are right. It just strikes me as odd that it does not tell you what is wrong. Only where something went wrong.

athei commented:
From my point of view all TODOs and comments are dealt with. Requested Review.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+
+        /// Add a new owner to the contract.
+        /// Panics if the owner already exists.
+        /// Only callable by the wallet itself.
```
The above might go too much into detail but it is a general doc-style preferred by the ink! codebase and Rust in general.
First line is short description, then empty line, followed by an optional longer description or note.
Also we make use of the general Rust docs-design principles to put `# Errors`, `# Panics`, `# Examples`, `# Note`, `# Safety` etc. sections in docs if needed.
Please apply to the other docs as well.

athei replied:
I have adjusted the documentation of the public items accordingly (and added an example).

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        fn add_owner(&mut self, new_owner: AccountId) {
+            self.ensure_from_wallet();
+            self.ensure_no_owner(&new_owner);
+            self.ensure_requirement_is_valid(self.owners.len() + 1, *self.requirement);
```
Btw.: I really like the defensive programming style used in this PR. :) Really good example of how contracts should be written.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        /// Remove an owner from the contract.
+        /// Only callable by the wallet itself. If by doing this the amount of owners
+        /// would be smaller than the requirement it is adjusted to be exactly the
+        /// number of owners. Panics if `owner` is no owner of the wallet.
```
Is this the semantics of the original multi-sig because I kinda dislike this implied effect but maybe I just don't know why this is useful to have. I'd generally just trap (panic) in such cases. If it is the original semantics we might just keep it as it is though.

athei replied:
Yes this is the gnosis semantic. But we can change it. We changed other things already.

athei replied:
I think we should leave it the way it is for the following reason: If we introduce the panic behavior you need to lower the requirement before removing an owner. This introduces a race where after lowering the requirement a transaction can get confirmed by too few people (including the owner you want to remove).

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+            self.ensure_from_wallet();
+            self.ensure_owner(&owner);
+            let len = self.owners.len() - 1;
+            let requirement = u32::min(len, *self.requirement.get());
```
```suggestion
            let requirement = u32::min(len, *self.requirement);
```

Robbepop replied:
Maybe I am wrong but it should generally work without the `.get()`

athei replied:
You are right.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+            .push_arg(&CallInput(&t.input))
+            .fire()
+            .map(|_| ())
+            .map_err(|_| ());
```
Have a hard time to believe that this is what `rustfmt` actually formatted this code to. 0.o

athei replied:
It did and the CI agrees that it's OK.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        /// This can be called by any owner.
+        /// Panics if `trans_id` is no valid transaction id.
+        #[ink(message)]
+        fn confirm_transaction(&mut self, trans_id: TransactionId) {
```
Maybe this could profit from a result that tells the caller if the transaction is ready for execution (has enough approvals). What do you think? This way you can more easily emulate the old-style of executing such a transaction right away. Or we could provide another method that does this job of the old semantics.

athei replied:
This makes sense. Now that the confirmation check is cheap this might be viable to return something useful here.

athei replied:
I added a `ConfirmationStatus`type that is returned from `submit_transaction` and `confirm_transaction`. It is also included in the confirmation event.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+
+        /// Execute a confirmed execution.
+        /// Its return type indicates whether the called transaction was succesful.
+        /// This can be called by anyone.
```
Does the original multi-sig return the result of the called function? Because this implementation doesn't but I think we could return some bytes back to the caller. This, however, should be another call since that's more work done. What do you think? Or should we keep it simple?

athei replied:
Ethereum contracts cannot return anything AFAIK. You completely rely on events there. Therefore you would detect the failure by listening for the callees events.

I think you are right. We should make use of return values as its possible in substrate.

If you put this in a separate call you must store the result in contract storage. I do not know if this would be so good.

athei replied:
I split `execute_transaction` into `eval_transaction` and `invoke_transaction`. The former one retrieves the output of the call as bytes.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        fn owner_index(&self, owner: &AccountId) -> u32 {
+            self.owners.iter().position(|x| *x == *owner).expect(
+                "This is only called after it was already verified that the id is
+                actually an owner.",
```
```suggestion
                 actually an owner.",
```

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+            );
+        }
+
+        /// Panic of the transaction `trans_id` does not exit.
```
```suggestion
        /// Panic if the transaction `trans_id` does not exit.
```

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+            .expect("User is responsible for only supplying existing keys.");
+        f(&mut count);
+        map.insert(*key, count);
+    }
```
Should we maybe have this in our `storage::BTreeMap` implementation? Is there a Rust `BTreeMap` equivalent

athei replied:
I think we should. I was surprised that there was no `mutate_with` like with he `HashMap`.

Robbepop replied:
But on the other handside it has an entry API.

athei replied:
I removed the function and switched to the entry API.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+
+        #[test]
+        #[should_panic]
+        fn add_owner_existing_fails() {
```
```suggestion
        fn add_existing_owner_fails() {
```

Robbepop replied:
Please no [Yoda](https://en.wikipedia.org/wiki/Yoda_conditions) identifiers. ;)

athei replied:
Comply I must.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        }
+
+        #[test]
+        fn cancel_transaction_nonexisting() {
```
So it doesn't panic when I cancel a non-existing transaction?

athei replied:
Yep. It just does nothing. The code is just a little simpler that way and uses less gas in the common case (at least I think it does).

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        #[test]
+        fn execute_transaction_works() {
+            // Execution of calls is currently unsupported in off-chain test.
+            // Calling execute_transaction panics in any case.
```
So sad. :(

athei replied:
Maybe we can change it from panic to "nothing at all". Then I can at least test if it tried to execute or failed for some other reason.

Robbepop replied:
We could try but on the other hand I don't think that this feature is even on the horizon.

athei replied:
I guess we leave it that way then and re-iterate on the example later.

athei commented on examples/multisig_plain/.cargo/config:
```diff
@@ -0,0 +1,4 @@
+[target.wasm32-unknown-unknown]
+rustflags = [
+	"-C", "link-args=-z stack-size=65536 --import-memory"
+]
```
Aye. It was created by `cargo contract`. Is that already fixed there?

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+
+use ink_lang as ink;
+
+#[ink::contract(version = "0.1.0", env = ink_core::env::DefaultEnvTypes)]
```
I know. I just wanted to leave that because there is no other documentation on how to specify an `EnvType`.

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+    pub struct Transaction {
+        /// The AccountId of the contract that is called in this transaction.
+        callee: AccountId,
+        /// The raw selector which is the function name of the `callee`that is called.
```
I think for an learner primitive types are certainly easier to grasp because there is no need to look them up. This was my motivation to choose them here.

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+
+    impl<'a> scale::Encode for CallInput<'a> {
+        fn encode_to<T: Output>(&self, dest: &mut T) {
+            dest.write(self.0);
```
Are you sure? My understanding is that `Output::write` only writes out the bytes and that `Output::push` would do the SCALE encoding of the slice.

Robbepop replied:
Ah I think you are right and I confused it with `push`. Sorry!

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+    struct MultisigPlain {
+        /// Every entry in this map represents the confirmation of an owner for a
+        /// transaction. This is effecively a set rather than a map.
+        confirmations: storage::BTreeMap<(TransactionId, AccountId), ()>,
```
I decided against that when implementing this contract for the following reason: You need stable owner indices for that. If you do enough owner removals additions and replacements you finally run out of indices and your contract is broken because you ran out of indices.

Maybe we can live with that because 64 ids is plenty but I wanted to opt for the safe approach in this example.

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        /// for referencing them in confirmation calls.
+        transactions: storage::Stash<Transaction>,
+        /// The list is a vector because iterating over it is necessary when cleaning
+        /// up the confirmation set.
```
I switched to `BTreeMap` which I cannot iterate either. Should I wait then for the iterator? Or iterate after merge?

Robbepop replied:
If even `BTreeMap` is missing iteration then just keep it as it is for now.

athei commented on examples/multisig_plain/lib.rs:
```diff
@@ -0,0 +1,837 @@
+        /// A list of owners must be supplied and a number of how many of them must
+        /// confirm a transaction. Duplicate owners are silently dropped.
+        #[ink(constructor)]
+        fn new(&mut self, owners: Vec<AccountId>, requirement: u32) {
```
Done.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -186,6 +193,17 @@ mod multisig_plain {
+        #[ink(topic)]
+        result: Result<Vec<u8>, ()>,
+    }
+
```
I think it would be more user friendly to have a single event for this that unifies both events.
E.g. convert `result` type to `Result<Option<Vec<u8>>, ()>` or something like that to signal if there was a return type or not.

athei replied:
Done.

Robbepop commented on examples/multisig_plain/lib.rs:
````diff
@@ -255,8 +260,36 @@ mod multisig_plain {
+        /// let (id, _) = submit_transaction(transaction);
+        /// // wait until all required owners have confirmed and then execute the transaction
+        /// invoke_transaction(id);
+        /// ```
````
nice

athei replied:
I reworked it to use `CallParams` to call into the contract. So it's less pseudo code and will actually compile once rustdocs will try to compile the example.

Robbepop commented on examples/multisig_plain/lib.rs:
```diff
@@ -339,8 +395,11 @@ mod multisig_plain {
-        /// Panics if `trans_id` is no valid transaction id.
+        ///
+        /// # Panics
+        /// If `trans_id` is no valid transaction id.
```
```suggestion
        /// # Panics
        ///
        /// If `trans_id` is no valid transaction id.
```

athei commented:
I rebased the branch onto the current master. I guess the force push removed your approval. Wondering if a merge would have done that, too. Should I avoid force pushing to PRs?

Robbepop commented:
> I rebased the branch onto the current master. I guess that removed your approval. Wondering if a merge would have done that, too. Should I avoid force pushing to PRs?

Since we squash merge in the end a merge wouldn't be too bad. ;)

Robbepop commented:
> > I rebased the branch onto the current master. I guess that removed your approval. Wondering if a merge would have done that, too. Should I avoid force pushing to PRs?
> 
> Since we squash merge in the end a merge wouldn't be too bad. ;)

I would have loved a CI output for this new example contract before merge but unfortunately CI is too broken. :( @TriplEight we really need a fix for this.

ascjones commented:
> I guess that removed your approval

Yeah this repo is configured to "dimiss" previous reviews when any more pushes are made to the PR branch, so requires a fresh approval

TriplEight commented:
These CI errors are not project-specific, restart of the job helps with them. I'm working on them, but right now have no clue what is the cause.

TriplEight commented:
Now, these examples failures are something you will want to fix.

athei commented:
> Now, these examples failures are something you will want to fix.

I am not sure what you are saying. Is there anything that I need to do right now?

TriplEight commented:
> > These CI errors are not project-specific, restart of the job helps with them. I'm working on > them, but right now have no clue what is the cause.
> 
> > Now, these examples failures are something you will want to fix.
> 
> I am not sure what you are saying. Is there anything that I need to do right now?

Nope, this kind of error is fixed in https://github.com/paritytech/ink/pull/343

athei commented:
Nice. I will rebase this branch onto this changes as soon as they hit master.

Robbepop commented:
Clippy warnings in this PR. We can merge once fixed.

athei commented:
Rebased on current master and fixed the clippy warning. CI still fails:
```
 error: failed to load source for a dependency on `derive_more`
Caused by:
  Unable to update registry `https://github.com/rust-lang/crates.io-index`
Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`
Caused by:
  missing delta bases; class=Indexer (15)
```

TriplEight commented:
@athei this is a different bug, I'm not sure what's the cause, but I'm on it. This is something random that happens on several projects CIs.
